File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/5. DP on Strings/javascript/05. Minimum steps to make string palindrome.js
/*
=============================================================================
  QUESTION: 1312. Min Insertions to Make String Palindrome (LeetCode)
=============================================================================

  Min insertions = n - LPS(s). Keep the longest palindromic subseq,
  insert mirror of remaining chars.

  Example: "mbadm" → 2 ("mbdabdm")

=============================================================================
  APPROACH: Space Optimized — O(N²) Time, O(N) Space
=============================================================================
*/

function minInsertions(s) {
    let r = s.split('').reverse().join('');
    let n = s.length;
    let prev = new Array(n + 1).fill(0);
    for (let i = 1; i <= n; i++) {
        let curr = new Array(n + 1).fill(0);
        for (let j = 1; j <= n; j++) {
            if (s[i - 1] === r[j - 1]) curr[j] = prev[j - 1] + 1;
            else curr[j] = Math.max(prev[j], curr[j - 1]);
        }
        prev = curr;
    }
    return n - prev[n]; // n - LPS
}

// ==========================================================================
console.log(minInsertions("zzazz"));  // 0
console.log(minInsertions("mbadm"));  // 2
console.log(minInsertions("leetcode"));  // 5

File: Strivers-A2Z-DSA-Sheet/14. Dynamic Programming/5. DP on Strings/python/05. Minimum steps to make string palindrome.py
"""05. Min insertions to make palindrome = n - LPS(s). Example: "mbadm" → 2"""

def min_insertions(s):
    r = s[::-1]; n = len(s)
    prev = [0]*(n+1)
    for i in range(1, n+1):
        curr = [0]*(n+1)
        for j in range(1, n+1):
            if s[i-1] == r[j-1]: curr[j] = prev[j-1] + 1
            else: curr[j] = max(prev[j], curr[j-1])
        prev = curr
    return n - prev[n]

if __name__ == "__main__":
    print(min_insertions("mbadm"))  # 2

